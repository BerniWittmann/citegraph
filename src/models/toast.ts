import uuid from 'uuid'

export enum ToastSeverity {
  SUCCESS = 'success',
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error'
}

export enum ToastIcon {
  SUCCESS = 'mdi-check-circle-outline',
  INFO = 'mdi-information-outline',
  WARNING = 'mdi-alert',
  ERROR = 'mdi-alert-circle-outline',
}

export enum ToastDelay {
  SUCCESS = 3000,
  INFO = 4000,
  WARNING = 5000,
  ERROR = 6000,
}

interface ToastData {
  severity: ToastSeverity;
  icon: ToastIcon;
  text: string;
  delay: number
}

export default class Toast implements ToastData {
  id: string;
  timestamp: number;
  severity: ToastSeverity;
  icon: ToastIcon;
  text: string;
  delay: number

  constructor ({ severity, icon, text, delay }: ToastData) {
    this.id = uuid.v4()
    this.timestamp = Date.now()
    this.severity = severity
    this.icon = icon
    this.text = text
    this.delay = delay
  }
}

export class SuccessToast extends Toast {
  constructor (text: string) {
    super({
      text,
      severity: ToastSeverity.SUCCESS,
      icon: ToastIcon.SUCCESS,
      delay: ToastDelay.SUCCESS
    })
  }
}

export class InfoToast extends Toast {
  constructor (text: string) {
    super({
      text,
      severity: ToastSeverity.INFO,
      icon: ToastIcon.INFO,
      delay: ToastDelay.INFO
    })
  }
}

export class WarningToast extends Toast {
  constructor (text: string) {
    super({
      text,
      severity: ToastSeverity.WARNING,
      icon: ToastIcon.WARNING,
      delay: ToastDelay.WARNING
    })
  }
}

export class ErrorToast extends Toast {
  constructor (text: string) {
    super({
      text,
      severity: ToastSeverity.ERROR,
      icon: ToastIcon.ERROR,
      delay: ToastDelay.ERROR
    })
  }
}
