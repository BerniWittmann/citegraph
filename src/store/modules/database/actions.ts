import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { DatabaseState } from '@/store/modules/database/types'
import Axios from 'axios'

export const actions: ActionTree<DatabaseState, RootState> = {
  async clearDatabase ({ dispatch }): Promise<undefined> {
    try {
      await Axios(`/database/clear`, {
        method: 'POST'
      })
      dispatch('toasts/showSuccess', 'settings.database.clear.success', { root: true })
      return undefined
    } catch (error) {
      console.error(error)
      dispatch('toasts/showError', 'settings.database.clear.error', { root: true })
      return error
    }
  },

  async reseedDatabase ({ dispatch }): Promise<undefined> {
    try {
      await Axios(`/database/reseed`, {
        method: 'POST'
      })
      dispatch('toasts/showSuccess', 'settings.database.reseed.success', { root: true })
      return undefined
    } catch (error) {
      console.error(error)
      dispatch('toasts/showError', 'settings.database.reseed.error', { root: true })
      return error
    }
  }
}
