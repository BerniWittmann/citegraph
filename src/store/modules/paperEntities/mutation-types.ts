export const SET_ENTITIES = 'SET_ENTITIES'
export const SET_ENTITY_TYPE = 'SET_ENTITY_TYPE'
export const SET_ENTITY_COUNT = 'SET_ENTITY_COUNT'
