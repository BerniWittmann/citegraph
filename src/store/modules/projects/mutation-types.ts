
export const SET_ACTIVE_PROJECT = 'SET_ACTIVE_PROJECT'
export const UNSET_ACTIVE_PROJECT = 'UNSET_ACTIVE_PROJECT'
export const SET_PROJECTS = 'SET_PROJECTS'
export const OPEN_PROJECT = 'OPEN_PROJECT'
export const CLOSE_PROJECT = 'CLOSE_PROJECT'
