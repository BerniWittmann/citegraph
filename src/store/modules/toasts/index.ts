import { Module } from 'vuex'
import { ToastsState } from './types'
import { RootState } from '@/store/types'
import { state } from './state'
import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'

const namespaced: boolean = true

export const toasts: Module<ToastsState, RootState> = {
  namespaced,
  state,
  getters,
  actions,
  mutations
}
