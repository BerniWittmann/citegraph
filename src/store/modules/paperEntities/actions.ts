import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { PaperEntitiesState } from './types'
import { PaperEntityQueryParameters } from '@/models/paperEntities/query'
import Axios from 'axios'
import * as types from '@/store/modules/paperEntities/mutation-types'
import jsLogger from 'js-logger'
import { entityKeysMap, PaperEntity, PaperEntityFields } from '@/models/paperEntities'
import { PaperEntityMutationParameters } from '@/models/paperEntities/mutation'

const logger = jsLogger.get('paperEntities/actions')

export const actions: ActionTree<PaperEntitiesState, RootState> = {
  async fetchEntities ({ commit, dispatch }, params: PaperEntityQueryParameters): Promise<undefined> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          query: EntityClass.getQuery(params)
        }
      })
      const data = response.data.data[EntityClass.queryName]
      const totalCount = data.count
      const payload: Array<PaperEntity> = EntityClass.transformer.fetchCollection(data[EntityClass.schemaName])
      commit(types.SET_ENTITIES, payload)
      commit(types.SET_ENTITY_COUNT, totalCount)
      commit(types.SET_ENTITY_TYPE, EntityClass.key)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.fetch_error', { root: true })
      commit(types.SET_ENTITIES, [])
      commit(types.SET_ENTITY_COUNT, -1)
      commit(types.SET_ENTITY_TYPE, undefined)
      return error
    }
  },

  async fetchEntity ({ commit, dispatch }, params: PaperEntityQueryParameters): Promise<void> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          query: EntityClass.getQuery(params)
        }
      })
      const data = response.data.data[EntityClass.queryName]
      const payload: PaperEntity = EntityClass.transformer.fetch(data[EntityClass.schemaName])
      commit(types.SET_ACTIVE_ENTITY, payload)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.fetch_single_error', { root: true })
      commit(types.SET_ACTIVE_ENTITY, undefined)
      return error
    }
  },

  async updateEntity ({ commit, dispatch }, {
    params,
    data
  }: { params: PaperEntityMutationParameters, data: Partial<PaperEntityFields>}): Promise<void> {
    try {
      const EntityClass = entityKeysMap[params.entityType]
      const response = await Axios(`/projects/${params.projectId}/paper-entities`, {
        method: 'POST',
        data: {
          mutation: EntityClass.getMutation(params, EntityClass.transformer.send(data))
        }
      })
      const responseData = response.data.data[EntityClass.queryName]
      const payload: PaperEntity = EntityClass.transformer.fetch(responseData)
      commit(types.UPDATE_ENTITY, payload)
      return undefined
    } catch (error) {
      logger.error(error)
      dispatch('toasts/showError', 'project.explore.update_single_error', { root: true })
      return error
    }
  }
}
