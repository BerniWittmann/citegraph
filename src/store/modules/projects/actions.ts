import Axios from 'axios'
import { ActionTree } from 'vuex'
import { ProjectsState } from '@/store/modules/projects/types'
import { RootState } from '@/store/types'
import Project from '@/models/project'
import * as types from './mutation-types'

export const actions: ActionTree<ProjectsState, RootState> = {
  async fetchProject ({ commit }, id: string): Promise<undefined> {
    try {
      const response = await Axios(`/projects/${id}`)
      const payload: Project = new Project(response && response.data)
      commit(types.SET_ACTIVE_PROJECT, payload)
      commit(types.OPEN_PROJECT, payload)
      return undefined
    } catch (error) {
      console.error(error)
      commit(types.UNSET_ACTIVE_PROJECT)
      return error
    }
  },

  unsetActiveProject ({ commit }): void {
    commit(types.UNSET_ACTIVE_PROJECT)
  },

  async fetchProjects ({ commit }): Promise<undefined> {
    try {
      const response = await Axios(`/projects`)
      const projects = response.data.map((project: Project) => new Project(project))
      commit(types.SET_PROJECTS, projects)
      return undefined
    } catch (error) {
      console.error(error)
      return error
    }
  },

  openProject ({ commit }, project: Project): void {
    commit(types.OPEN_PROJECT, project)
  },

  closeProject ({ commit }, project: Project): void {
    commit(types.CLOSE_PROJECT, project)
  },

  async createProject ({ commit }, data: Project): Promise<Project | undefined> {
    try {
      const response = await Axios(`/projects`, {
        method: 'POST',
        data: data
      })
      const project = new Project(response.data)
      commit(types.ADD_PROJECT, project)
      return project
    } catch (error) {
      console.error(error)
      return error
    }
  }
}
