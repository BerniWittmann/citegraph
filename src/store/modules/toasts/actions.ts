import { ActionTree } from 'vuex'
import { RootState } from '@/store/types'
import { ToastsState } from '@/store/modules/toasts/types'
import * as types from '@/store/modules/toasts/mutation-types'
import Toast, { ErrorToast, InfoToast, SuccessToast, WarningToast } from '@/models/toast'

export const actions: ActionTree<ToastsState, RootState> = {
  showSuccess ({ commit }, message: string): void {
    commit(types.ADD_TOAST, new SuccessToast(message))
  },

  showInfo ({ commit }, message: string): void {
    commit(types.ADD_TOAST, new InfoToast(message))
  },

  showWarning ({ commit }, message: string): void {
    commit(types.ADD_TOAST, new WarningToast(message))
  },

  showError ({ commit }, message: string): void {
    commit(types.ADD_TOAST, new ErrorToast(message))
  },

  removeToast ({ commit }, toast: Toast): void {
    commit(types.REMOVE_TOAST, toast)
  }
}
