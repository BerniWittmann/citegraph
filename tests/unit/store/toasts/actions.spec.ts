import { actions } from '@/store/modules/toasts/actions'
import * as mutationTypes from '@/store/modules/toasts/mutation-types'
import { ErrorToast, InfoToast, SuccessToast, WarningToast } from '@/models/toast'

describe('store/modules/toasts/actions', () => {
  it('provides the actions', () => {
    expect(actions).toMatchSnapshot()
  })

  describe('showSuccess', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.showSuccess as Function
      action({ commit }, 'My message')
      expect(commit).toHaveBeenCalledWith(mutationTypes.ADD_TOAST, expect.any(SuccessToast))
      const toast = commit.mock.calls[0][1]
      expect(toast.text).toEqual('My message')
      expect(toast.severity).toEqual('success')
    })
  })

  describe('showInfo', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.showInfo as Function
      action({ commit }, 'My message')
      expect(commit).toHaveBeenCalledWith(mutationTypes.ADD_TOAST, expect.any(InfoToast))
      const toast = commit.mock.calls[0][1]
      expect(toast.text).toEqual('My message')
      expect(toast.severity).toEqual('info')
    })
  })

  describe('showWarning', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.showWarning as Function
      action({ commit }, 'My message')
      expect(commit).toHaveBeenCalledWith(mutationTypes.ADD_TOAST, expect.any(WarningToast))
      const toast = commit.mock.calls[0][1]
      expect(toast.text).toEqual('My message')
      expect(toast.severity).toEqual('warning')
    })
  })

  describe('showError', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.showError as Function
      action({ commit }, 'My message')
      expect(commit).toHaveBeenCalledWith(mutationTypes.ADD_TOAST, expect.any(ErrorToast))
      const toast = commit.mock.calls[0][1]
      expect(toast.text).toEqual('My message')
      expect(toast.severity).toEqual('error')
    })
  })

  describe('removeToast', () => {
    it('triggers the mutation', () => {
      const commit = jest.fn()
      const action = actions.removeToast as Function
      const toast = new SuccessToast('My message')
      action({ commit }, toast)
      expect(commit).toHaveBeenCalledWith(mutationTypes.REMOVE_TOAST, toast)
    })
  })
})
