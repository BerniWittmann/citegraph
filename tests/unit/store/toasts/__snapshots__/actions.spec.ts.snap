// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store/modules/toasts/actions provides the actions 1`] = `
Object {
  "removeToast": [Function],
  "showError": [Function],
  "showInfo": [Function],
  "showSuccess": [Function],
  "showWarning": [Function],
}
`;
